File: lib/theme/typography.dart
import 'package:flutter/widgets.dart';

/// Type system: Instrument Serif (display) + Archivo (UI).
///
/// Both families are bundled in `pubspec.yaml`, so a weight is selected with a
/// real font file rather than being synthesised.
abstract final class FontFamily {
  static const display = 'InstrumentSerif';
  static const ui = 'Archivo';
}

/// Archivo at [size] / [weight], with optional tracking and line height.
///
/// `height` in Flutter is a multiplier of the font size, which is exactly what
/// the CSS `line-height: 1.3` values in the handoff mean.
TextStyle ui(
  double size, {
  FontWeight weight = FontWeight.w400,
  Color? color,
  double? height,
  double? letterSpacing,
}) {
  return TextStyle(
    fontFamily: FontFamily.ui,
    fontSize: size,
    fontWeight: weight,
    color: color,
    height: height,
    letterSpacing: letterSpacing,
  );
}

/// Instrument Serif at [size], with the design's line-height/tracking pairs.
TextStyle display(
  double size, {
  Color? color,
  double? height,
  double? letterSpacing,
}) {
  return TextStyle(
    fontFamily: FontFamily.display,
    fontSize: size,
    fontWeight: FontWeight.w400,
    color: color,
    height: height,
    letterSpacing: letterSpacing,
  );
}

/// The overline used for every small uppercase label in the design
/// (date, sheet titles): 11pt / 700 / 0.16em / uppercase.
///
/// The text itself is uppercased at the call site — Flutter has no
/// `text-transform`, and Turkish casing needs the locale-aware helper anyway.
TextStyle overline({Color? color}) => ui(
      11,
      weight: FontWeight.w700,
      color: color,
      letterSpacing: 11 * 0.16,
    );

File: src/theme/typography.ts
import { TextStyle } from 'react-native';

/**
 * Type system: Instrument Serif (display) + Archivo (UI).
 *
 * React Native cannot synthesise weights for bundled fonts, so a weight is
 * selected by picking the right family file rather than by `fontWeight`.
 */

export const fontFamily = {
  display: 'InstrumentSerif_400Regular',
  ui400: 'Archivo_400Regular',
  ui500: 'Archivo_500Medium',
  ui600: 'Archivo_600SemiBold',
  ui700: 'Archivo_700Bold',
  ui800: 'Archivo_800ExtraBold',
} as const;

export type UIWeight = 400 | 500 | 600 | 700 | 800;

const UI_FAMILY: Record<UIWeight, string> = {
  400: fontFamily.ui400,
  500: fontFamily.ui500,
  600: fontFamily.ui600,
  700: fontFamily.ui700,
  800: fontFamily.ui800,
};

/** Archivo at `size` / `weight`. */
export function ui(size: number, weight: UIWeight = 400): TextStyle {
  return { fontFamily: UI_FAMILY[weight], fontSize: size, fontWeight: undefined };
}

/** Instrument Serif at `size`, with the design's line-height/tracking pairs. */
export function display(size: number, lineHeight?: number, letterSpacing?: number): TextStyle {
  return {
    fontFamily: fontFamily.display,
    fontSize: size,
    ...(lineHeight !== undefined ? { lineHeight } : null),
    ...(letterSpacing !== undefined ? { letterSpacing } : null),
  };
}

/**
 * Overline used for every small uppercase label in the design
 * (date, sheet titles): 11pt / 700 / 0.16em / uppercase.
 */
export const overline: TextStyle = {
  ...ui(11, 700),
  letterSpacing: 11 * 0.16,
  textTransform: 'uppercase',
};
